### PR TITLE
 integrated with inbuilt key manager

### PR DESCRIPTION
**Pull request by ruks**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

-  integrated with inbuilt key manager

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/RestCallUtil.java [hunk 1]
```diff
@@ -90,11 +91,12 @@ HttpResponse getRequest(URI uri, MediaType acceptContentType, List<String> cooki
      * @param cookies            List of cookies to be added with the request, if any
      * @param entity             Entity which needs to be sent to the service with the POST request
      * @param payloadContentType MediaType of the payload added in the request
+     * @param headers            additional header key/value map
      * @return HttpResponse from service
      * @throws APIManagementException In case of any failures, when trying to make a POST request
      */
     HttpResponse postRequest(URI uri, MediaType acceptContentType, List<String> cookies, Entity entity,
-                             MediaType payloadContentType) throws APIManagementException;
+            MediaType payloadContentType, Map<String, String> headers) throws APIManagementException;
```
Comment (pubudu538) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/RestCallUtil.java]: Please check for the formatting issues.

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/LightWeightKeyManagerImpl.java [hunk 2]
```diff
@@ -0,0 +1,469 @@
+package org.wso2.carbon.apimgt.core.impl;
```
Comment (pubudu538) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/LightWeightKeyManagerImpl.java]: Shall we change the name of the class? I guess we don't need to explicitly say about lightweight key manager.

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/LightWeightKeyManagerImpl.java [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+    }
+
+    public LightWeightKeyManagerImpl(KeyMgtConfigurations keyManagerConfigs, RestCallUtil restCallUtil) {
+        this.keyManagerConfigs = keyManagerConfigs;
+        this.restCallUtil = restCallUtil;
+    }
+
+    @Override
+    public OAuthApplicationInfo createApplication(OAuthAppRequest oauthAppRequest) throws KeyManagementException {
+        if (log.isDebugEnabled()) {
```
Comment (pubudu538) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/LightWeightKeyManagerImpl.java]: We can use https://github.com/wso2/carbon-apimgt/blob/master/components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/util/APIUtils.java#L65

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/LightWeightKeyManagerImpl.java [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+                            + ". Status Code: " + httpResponse.getResponseCode(),
+                    ExceptionCodes.OAUTH2_APP_CREATION_FAILED);
+        } else {  //Unknown Error
+            throw new KeyManagementException(
+                    "Error occurred while DCR application creation. Error: " + " Status Code: " + httpResponse
+                            .getResponseCode(), ExceptionCodes.OAUTH2_APP_CREATION_FAILED);
+        }
+    }
+
+    @Override
```
Comment (pubudu538) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/LightWeightKeyManagerImpl.java]: Are we not adding @see annotation for Interface method comment? @harsha89

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/RestCallUtilImpl.java [hunk 5]
```diff
@@ -203,8 +203,8 @@ public HttpResponse getRequest(URI uri, MediaType acceptContentType, List<String
      * {@inheritDoc}
      */
     @Override
-    public HttpResponse postRequest(URI uri, MediaType acceptContentType, List<String> cookies,
-                                    Entity entity, MediaType payloadContentType) throws APIManagementException {
+    public HttpResponse postRequest(URI uri, MediaType acceptContentType, List<String> cookies, Entity entity,
+            MediaType payloadContentType, Map<String, String> headers) throws APIManagementException {
```
Comment (pubudu538) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/RestCallUtilImpl.java]: Formatting issues?

Comment (ruks) [reply components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/RestCallUtilImpl.java]: checked the format and it is fine

**Pull request merged**